### DISCUSSION TITLE
Google Sheet support

### DISCUSSION THREAD

**Discussion by mantielero** [arendst/Tasmota#15441] [Ideas]
I think it would be nice being able to upload data to Google Sheet. I've been experimenting myself with `WebQuery` (but no success yet).

Using WebQuery it should be something like this (as I said, not working yet):
```
WebQuery https://sheets.googleapis.com/v4/spreadsheets/<spreadsheet_id>/values/<range>:append?valueInputOption=RAW POST [Authorization: Bearer <REFRESHER_ACCESS_CODE>|Content-Type: application/json] {"values": [[9999]]} 
```

Using `curl` is working for me with the following:
```
curl -X POST \
   https://sheets.googleapis.com/v4/spreadsheets/<spreadsheet_id>/values/<range>:append?valueInputOption=RAW \
  -H "Authorization: Bearer <REFRESHER_ACCESS_CODE>" \
  -H "Content-Type: application/json" \
  -d '{"values": [[9999]]}'
```

**Comment by barbudor** [comment]
Are you on ESP32 ?
AFAIK https is not supported on ESP8266

If on ESP32 you may find it more flexbible to use Berry and its WebClient class

**Reply by mantielero** [reply]
I am in ESP8266. That might be the reason behind the:
```
RESULT = {"WebQuery":"Connect failed"}
```

Thanks!

**Reply by barbudor** [reply]
My opinion is that such things should be left to the back-end which in my case is Node-Red
Like I don't use the InfluxDB driver in Tasmota but use Node-Red to write the data from Tasmota sensors to the database

But that's just my opinion 😁

**Reply by mantielero** [reply]
I am new to tasmota (really impressed with it to be honest). I come from a RPi4 running OpenHAB, but I really like having a little micro doing so much consuming so little energy and no moving parts. I just use it to receive data from 3 Zigbee sensors (2 Temperature, 1 door) and send data via Telegram. When the door opens I get an alert, and when I send `Event T`, I get the temperatures, humidities and pressures from the sensors.

Now I am exploring how to upload the data somewhere online. I would prefer to avoid having another box running 24h at home. So any suggestion is welcomed.

**Reply by sfromis** [reply]
On ESP8266, you'd have to use http instead of https, and I'd certainly not expect Google to accept unencrypted http.

**Comment by mantielero** [comment]
Just for the record, I managed to both read and write data from Excel using berry as suggested and the ESP32.

Reading:
```berry
cl = webclient()
cl.begin("https://sheets.googleapis.com/v4/spreadsheets/<SPREADSHEET_ID>/values/<SHEET_ID>")
cl.add_header("Authorization","Bearer <BEARER_ACCESS_CODE>")
#cl.add_header("Content-Type","application/json")
cl.GET()
print(cl.get_string())
```

Writting:
```berry
cl = webclient()
cl.begin("https://sheets.googleapis.com/v4/spreadsheets/<SPREADSHEET_ID>/values/<SHEET_ID>:append?valueInputOption=RAW")
cl.add_header("Authorization","Bearer <BEARER_ACCESS_CODE>")
cl.add_header("Content-Type","application/json")
cl.POST("{\"values\": [[12345,6789]]}")
```
Things get a bit more complicated with the OAUTH2 logic (refresher codes, bearer code, ...).

**Reply by s-hadinger** [reply]
Well done. Thanks for sharing.

**Comment by mantielero** [comment]
It looks like another (somehow easier approach) would be to support google script. The main advantage is that you don't need oauth2 at all. It could be used like this:
```
WebQuery https://script.google.com/macros/s/<script_id>/exec?temperature=12.3&humidity=45.3&pressure=980.3
```
The following function in google script would populate the Google Sheet with the data. There are many [tutorials](https://circuitdigest.com/microcontroller-projects/log-temperature-sensor-data-to-google-sheet-using-nodemcu-esp8266) about how to create web app given a function like the this (easy even for my IT skills):
```gs
var ssId = "<spreadsheet_id>";
var sName = "<sheet_name>";

function doGet(e) {
  var temperature = e.parameter.temperature;
  var humidity = e.parameter.humidity;
  var pressure = e.parameter.pressure;

  var ss = SpreadsheetApp.openById(ssId);
  var s = ss.getSheetByName(sName) ? ss.getSheetByName(sName) : ss.insertSheet(sName);

  s.appendRow([new Date(), temperature, humidity, pressure]);

  return HtmlService.createHtmlOutput("ok"); // + JSON.stringify(e));  
}
```

What would be needed? It would be nice having something similar to the Telegram module. Enabling to set the script ID and having a command such as `gsSend temperature=12.3&humidity=45.3&pressure=980.3`. A big plus would be if it worked also in the ESP8266, like Telegram does. I'm not sure if [HTTPSRedirect](https://github.com/jbuszkie/HTTPSRedirect) might be useful for this.

**Reply by sfromis** [reply]
Sounds like what you aim for has similarities to the [support for InfluxDB](https://github.com/arendst/Tasmota/blob/development/tasmota/xdrv_59_influxdb.ino#L21-L48).

**Reply by mantielero** [reply]
Yes it does. I haven't tried InfluxDB yet. 

But the Google Script approach would give me more flexibility. For instance, I would implement the Telegram integration bit in Google Script rather than in the ESP.

**Reply by barbudor** [reply]
Or Node-red 🤔😂

**Reply by sfromis** [reply]
Me pointing at the InfluxDB integration was in no way meant as a suggestion to use InfluxDB, but a reaction to your musings about what would be needed to integrate Google Sheet/Script support in Tasmota. The InfluxDB support has significant parallels, which the Telegram support does not in the same way.

**Reply by mantielero** [reply]
Ahh sorry. I misunderstood it.
